Docs setup for the Task SDK (#51153)
mistercrunch/Airflow: task-sdk/docs/api.rst

 .. Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

airflow.sdk API Reference
=========================

> This page documents the full public API exposed in Airflow 3.0+ via the Task SDK python module.

Reviewer: ```suggestion
This page documents the full public API exposed in Airflow 3 and later via the Task SDK python module.
```

Author: Changed it.

If something is not on this page it is best to assume that it is not part of the public API and use of it is entirely at your own risk

> -- we won't go out of our way break usage of them, but we make no promises either.

Reviewer: ```suggestion
-- we won't go out of our way and break usage of them, but we make no promises either otherwise.
```

Author: Changed it.

Defining DAGs
-------------
.. autoapiclass:: airflow.sdk.DAG

> .. autoapifunction:: airflow.sdk.dag

Reviewer: This is the dag decorator right? Either move it below into "Decorators" and remove from here or copy it there too.

Author: Moved it to decorator.

Decorators
----------
.. autoapifunction:: airflow.sdk.task

.. autoapifunction:: airflow.sdk.task_group

.. autoapifunction:: airflow.sdk.setup

.. autoapifunction:: airflow.sdk.teardown

.. autofunction:: airflow.sdk.task
.. autofunction:: airflow.sdk.setup
.. autofunction:: airflow.sdk.teardown
.. autofunction:: airflow.sdk.asset

Bases
-----

> .. autoapiclass:: airflow.sdk.BaseOperator

Reviewer: We are missing `BaseXCom` here.

Author: Added it

If something is not on this page it is best to assume that it is not part of the public API and use of it is entirely at your own risk
-- we won't go out of our way break usage of them, but we make no promises either.

Defining DAGs
-------------
.. autoapiclass:: airflow.sdk.DAG

.. autoapifunction:: airflow.sdk.dag

Decorators
----------
.. autoapifunction:: airflow.sdk.task

.. autoapifunction:: airflow.sdk.task_group

.. autoapifunction:: airflow.sdk.setup

.. autoapifunction:: airflow.sdk.teardown

.. autofunction:: airflow.sdk.task
.. autofunction:: airflow.sdk.setup
.. autofunction:: airflow.sdk.teardown
.. autofunction:: airflow.sdk.asset

Bases
-----
.. autoapiclass:: airflow.sdk.BaseOperator

.. autoapiclass:: airflow.sdk.BaseSensorOperator

.. autoapiclass:: airflow.sdk.BaseNotifier

.. autoapiclass:: airflow.sdk.BaseOperatorLink

.. autoapiclass:: airflow.sdk.XComArg

.. autoapiclass:: airflow.sdk.PokeReturnValue

Connections & Variables
-----------------------
.. autoapiclass:: airflow.sdk.Connection

.. autoapiclass:: airflow.sdk.Variable

Tasks & Operators
-----------------
.. autoapiclass:: airflow.sdk.TaskGroup

> .. autoapiclass:: airflow.sdk.MappedOperator

Reviewer 2: Should nuke MappedOperator

Author: Yes. pushed just now

 .. Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

airflow.sdk API Reference
=========================

This page documents the full public API exposed in Airflow 3.0+ via the Task SDK python module.

If something is not on this page it is best to assume that it is not part of the public API and use of it is entirely at your own risk
-- we won't go out of our way break usage of them, but we make no promises either.

> .. :py:module: airflow.sdk
> 

Reviewer 3: This is a comment right?
```suggestion
```

Author: Yes.. removed it.

.. autoapifunction:: airflow.sdk.task_group

.. autoapifunction:: airflow.sdk.setup

.. autoapifunction:: airflow.sdk.teardown

.. autofunction:: airflow.sdk.task
.. autofunction:: airflow.sdk.setup
.. autofunction:: airflow.sdk.teardown
.. autofunction:: airflow.sdk.asset

Bases
-----
.. autoapiclass:: airflow.sdk.BaseOperator

.. autoapiclass:: airflow.sdk.BaseSensorOperator

.. autoapiclass:: airflow.sdk.BaseNotifier

.. autoapiclass:: airflow.sdk.BaseOperatorLink

.. autoapiclass:: airflow.sdk.XComArg

.. autoapiclass:: airflow.sdk.PokeReturnValue

Connections & Variables
-----------------------
.. autoapiclass:: airflow.sdk.Connection

.. autoapiclass:: airflow.sdk.Variable

Tasks & Operators
-----------------
.. autoapiclass:: airflow.sdk.TaskGroup

.. autoapiclass:: airflow.sdk.MappedOperator

.. autoapifunction:: airflow.sdk.get_current_context

.. autoapifunction:: airflow.sdk.get_parsing_context

.. autoapiclass:: airflow.sdk.Param

Setting Dependencies
~~~~~~~~~~~~~~~~~~~~
.. autoapifunction:: airflow.sdk.chain

.. autoapifunction:: airflow.sdk.chain_linear

.. autoapifunction:: airflow.sdk.cross_downstream

.. autoapifunction:: airflow.sdk.literal

Edges & Labels

> --------------

Reviewer 3: ```suggestion
~~~~~~~~~~~~~~
```

so that it's a sub heading under Task & Ops? (I think that makes sense from a layout pov? Not 100% sure)

Author: This makes sense.

    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

airflow.sdk API Reference
=========================

This page documents the full public API exposed in Airflow 3.0+ via the Task SDK python module.

If something is not on this page it is best to assume that it is not part of the public API and use of it is entirely at your own risk
-- we won't go out of our way break usage of them, but we make no promises either.

.. :py:module: airflow.sdk

Defining DAGs
-------------
.. autoapiclass:: airflow.sdk.DAG

.. autoapifunction:: airflow.sdk.dag

Decorators
----------
.. autoapifunction:: airflow.sdk.task

.. autoapifunction:: airflow.sdk.task_group

.. autoapifunction:: airflow.sdk.setup

.. autoapifunction:: airflow.sdk.teardown

> .. autofunction:: airflow.sdk.task

Reviewer 3: Future work: we need to document the "sub" decorators such as `@task.skip_if` etc.

Author: I can create subsequent PRs for this.

Author: https://github.com/apache/airflow/issues/51518

.. autofunction:: airflow.sdk.setup

> .. autofunction:: airflow.sdk.teardown

Reviewer 3: Future work: these are missing doc strings or any indication of the params they take.

Author: I can create subsequent PRs for this.

Author: https://github.com/apache/airflow/issues/51518

.. autofunction:: airflow.sdk.asset

Utilities
---------
.. autoapifunction:: airflow.sdk.chain

.. autoapifunction:: airflow.sdk.chain_linear

.. autoapifunction:: airflow.sdk.cross_downstream

> .. autoapifunction:: airflow.sdk.literal

Reviewer 3: I'd put these "tasks and operators", possibly under a sub section of "setting dependencies" or something

Author: Moved it

.. autoapifunction:: airflow.sdk.__version__

> .. autofunction:: airflow.sdk.__version__

Reviewer 3: a) Why do we have this twice?
b) it looks wrong:
<img width="1261" alt="Screenshot 2025-06-03 at 10 27 13" src="https://github.com/user-attachments/assets/0ef7163a-5b18-4607-b270-cea43ab3da88" />

c) do we even want to document this? I think not.

Author: a. `autoapifunction` and `autofunction` aren’t actually the same Sphinx directive. In  conf.py, I am loading both
- autoapi = static signature + link into the separate API‐tree
- autodoc = full, live doc‐string detail in place.

b. It looks wrong. I have removed it

Author: c. I was thinking of showing the version

Reviewer 3: The rendered docs  for publishing will show the version selector

<img width="1122" alt="Screenshot 2025-06-03 at 14 24 54" src="https://github.com/user-attachments/assets/e9b3d5c6-e902-4b63-b2d7-3ace7097917a" />

so we don't need to try and show the version here.

Author: Makes sense.

Bases
-----
.. autoapiclass:: airflow.sdk.BaseOperator

.. autoapiclass:: airflow.sdk.BaseSensorOperator

.. autoapiclass:: airflow.sdk.BaseNotifier

.. autoapiclass:: airflow.sdk.BaseOperatorLink

.. autoapiclass:: airflow.sdk.XComArg

.. autoapiclass:: airflow.sdk.PokeReturnValue

Connections & Variables
-----------------------
.. autoapiclass:: airflow.sdk.Connection

.. autoapiclass:: airflow.sdk.Variable

Tasks & Operators
-----------------
.. autoapiclass:: airflow.sdk.TaskGroup

.. autoapiclass:: airflow.sdk.MappedOperator

.. autoapifunction:: airflow.sdk.get_current_context

.. autoapifunction:: airflow.sdk.get_parsing_context

.. autoapiclass:: airflow.sdk.Param

Edges & Labels
--------------
.. autoapiclass:: airflow.sdk.EdgeModifier

.. autoapiclass:: airflow.sdk.Label

Metadata
--------

> .. autoapiclass:: airflow.sdk.Metadata

Reviewer 3: This should be part of assets I think, given it's doc string is `Metadata to attach to an AssetEvent.`

Author: Moved it to assets

Assets
------
.. autoapiclass:: airflow.sdk.Asset

.. autoapiclass:: airflow.sdk.AssetAlias

.. autoapiclass:: airflow.sdk.AssetAll

.. autoapiclass:: airflow.sdk.AssetAny

.. autoapiclass:: airflow.sdk.AssetWatcher

I/O Helpers
-----------
.. autoapiclass:: airflow.sdk.ObjectStoragePath

Execution Time Components
-------------------------
.. rubric:: Supervisor

.. autoapimodule:: airflow.sdk.execution_time.supervisor
   :members:
   :undoc-members:

.. rubric:: TaskRunner

.. autoapimodule:: airflow.sdk.execution_time.task_runner
   :members:

>    :undoc-members:

Reviewer 3: These explicitly must not be documented -- they are not part of the public interface.

Author: Removed it

.. rubric:: Context
.. autoapiclass:: airflow.sdk.Context
.. autoapimodule:: airflow.sdk.execution_time.context
   :members:
   :undoc-members:

.. rubric:: Cache

.. autoapiclass:: airflow.sdk.SecretCache

> .. autoclass:: airflow.sdk.SecretCache

Reviewer 3: Internal impl detail, don't document

Author: mypy-airflow-core was failing. That's the reason I added it: https://github.com/apache/airflow/actions/runs/15343037924/job/43173553034?pr=51153

Author: I've exposed MappedOperator in the SDK exactly like SecretCache so that both runtime imports and mypy picks it up

Reviewer 3: Mypy is an entirely different issue to what appears in the rendered docs.

Reviewer 3: And this error:

```

airflow-core/src/airflow/models/variable.py:34: error: Module "airflow.sdk" has
no attribute "SecretCache"  [attr-defined]
    from airflow.sdk import SecretCache
    ^
```

also means we should not import it from `airflow.sdk` directly, but from `airflow.sdk.execution_time...` etc.

Author: Removed it

API Client
----------
.. autoapimodule:: airflow.sdk.api.client
   :members:

>    :undoc-members:

Reviewer 3: ```suggestion
```

Author: removed it

    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

airflow.sdk API Reference
=========================

This page documents the full public API exposed in Airflow 3.0+ via the Task SDK python module.

If something is not on this page it is best to assume that it is not part of the public API and use of it is entirely at your own risk
-- we won't go out of our way break usage of them, but we make no promises either.

.. :py:module: airflow.sdk

Defining DAGs
-------------

.. autoapiclass:: airflow.sdk.DAG

.. autoapifunction:: airflow.sdk.dag

.. autoapiclass:: airflow.sdk.TaskGroup

.. autoapifunction:: airflow.sdk.get_parsing_context

.. autoapiclass:: airflow.sdk.definitions.context.AirflowParsingContext
   :undoc-members:
   :members:

> Tasks and Operators

Reviewer 2: We should also have  `task` decorator, `Param`, `setup`, `teardown` etc

Author: I have added a section for this.
